### PR TITLE
refactor(backends): split memtable existence check out

### PR DESCRIPTION
**Pull request by cpcloud**
First of ~two~ a few PRs to implement #10044. This one factors out memtable existence checks to allow a generic implementation of register-only-if-not-exists, as well as to allow improvement over time of this check as we discover faster existence check implementations.

### CODE REVIEW

File: ibis/backends/druid/__init__.py [hunk 1]
```diff
@@ -33,7 +33,6 @@ class Backend(SQLBackend):
     name = "druid"
     compiler = sc.druid.compiler
     supports_create_or_replace = False
-    supports_in_memory_tables = True
```
Comment (cpcloud) [inline ibis/backends/druid/__init__.py:36]: The default value of this flag is `True`, hence the removal in a few places.

File: ibis/backends/exasol/__init__.py [hunk 2]
```diff
@@ -253,50 +253,47 @@ def _register_in_memory_table(self, op: ops.InMemoryTable) -> None:
                 f"got null typed columns: {null_columns}"
             )
 
-        # only register if we haven't already done so
-        if (name := op.name) not in self.list_tables():
```
Comment (cpcloud) [inline ibis/backends/exasol/__init__.py:257]: The only bits I changed here were to remove this conditional, and then dedent what was previously indented inside the branch.

**Comment by cpcloud** [comment]
I'll split out the perf improvements to a separate PR.

File: ibis/backends/__init__.py [hunk 3]
```diff
@@ -1109,11 +1109,15 @@ def _register_udfs(self, expr: ir.Expr) -> None:
         if self.supports_python_udfs:
             raise NotImplementedError(self.name)
 
+    def _table_exists(self, name: str) -> bool:
```
Comment (cpcloud) [inline ibis/backends/__init__.py:1112]: Perhaps this should be call `_memtable_exists` to discourage expand the scope to include `database`/`catalog` lookups.

Comment (jcrist) [reply ibis/backends/__init__.py:1112]: You're too fast for me - I had this exact comment written in a pending review :). For consistency with the other methods it should probably be `_in_memory_table_exists` though (or rename the other ones :shrug:).

**Review by jcrist — APPROVED** [review]
LGTM (besides the method name).

File: ibis/backends/__init__.py (hunk 3 above)
Comment (cpcloud) [reply ibis/backends/__init__.py]: Renamed to `_in_memory_table_exists`.

**Pull request merged**